ykrt/src/location.rs

```diff
@@ -97,88 +29,407 @@ impl State {
+    //   <count>       | 1           | 0         | 0         | Start state
+    //                 | 1           | 0         | 1         | Illegal state
+    //                 | 1           | 1         | 0         | Illegal state
+    //                 | 1           | 1         | 1         | Illegal state
+    //   <HotLocation> | 0           | 0         | 0         | Not locked, no-one waiting
+    //   <HotLocation> | 0           | 0         | 1         | Locked, no thread(s) waiting
+    //   <HotLocation> | 0           | 1         | 0         | Not locked, thread(s) waiting
+    //   <HotLocation> | 0           | 1         | 1         | Locked, thread(s) waiting
+    //
+    // where `<count>` is an integer and `<HotLocation>` is a boxed `HotLocation` enum.
+    //
+    // The precise semantics of locking and, in particular, parking are subtle: interested readers
+    // are directed to https://github.com/Amanieu/parking_lot/blob/master/src/raw_mutex.rs#L33 for
+    // a more precise definition.
     state: AtomicUsize,
     phantom: PhantomData<I>,
 }
 
 impl<I> Location<I> {
+    /// Create a new location.
     pub fn new() -> Self {
+        // Locations start in the counting state with a count of 0.
         Self {
-            state: AtomicUsize::new(PHASE_COUNTING),
+            state: AtomicUsize::new(State::<I>::new().x),
             phantom: PhantomData,
         }
     }
 
-    pub(crate) fn load(&self, order: Ordering) -> State {
-        State(self.state.load(order))
+    /// Return this Location's internal state.
+    pub(crate) fn load(&self, order: Ordering) -> State<I> {
+        State {
+            x: self.state.load(order),
+            marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn compare_exchange_weak(
+        &self,
+        current: State<I>,
+        new: State<I>,
+        success: Ordering,
+        failure: Ordering,
+    ) -> Result<State<I>, State<I>> {
+        match self
+            .state
+            .compare_exchange_weak(current.x, new.x, success, failure)
+        {
+            Ok(x) => Ok(State {
+                x,
+                marker: PhantomData,
+            }),
+            Err(x) => Err(State {
+                x,
+                marker: PhantomData,
+            }),
+        }
+    }
+
+    /// Locks this `State` with `Acquire` ordering. If the location moves to the Counting state
+    /// during execution, this will return `Err`.
+    pub(crate) fn lock(&self) -> Result<State<I>, ()> {
+        {
+            let ls = self.load(Ordering::Relaxed);
+            debug_assert!(!ls.is_counting());
+            let new_ls = ls.with_lock().with_unparked();
+            if self
+                .state
+                .compare_exchange_weak(
+                    ls.with_unlock().with_unparked().x,
+                    new_ls.x,
+                    Ordering::Acquire,
+                    Ordering::Relaxed,
```

Comment:
It seems you came to the same conclusion I did about the second ordering. If it fails, we don't care what ordering we use.

Perhaps there's a niche use-case where you do care, but ours isn't one.

Reply:
Correct (`parking_lot` does the same thing for the same reason).